jbpm-runtime-manager/src/main/java/org/jbpm/runtime/manager/impl/PerCaseRuntimeManager.java

```diff
@@ -116,66 +115,31 @@ public RuntimeEngine getRuntimeEngine(Context<?> context) {
         if (!(context instanceof ProcessInstanceIdContext || context instanceof CaseContext)) {
             logger.warn("ProcessInstanceIdContext or CaseContext shall be used when interacting with PerCase runtime manager");
         }
-        
-        if (engineInitEager) {
-            KieSession ksession = null;
-            Long ksessionId = null;
 
-            RuntimeEngine localRuntime = findLocalRuntime(contextId);
-            if (localRuntime != null) {
-                return localRuntime;
-            }
-            synchronized (this) {
-                ksessionId = mapper.findMapping(context, this.identifier);
-                if (ksessionId == null) {
-                    ksession = factory.newKieSession();
-                    ksessionId = ksession.getIdentifier();
-
-                    if (context instanceof CaseContext) {
-                        ksession.execute(new SaveMappingCommand(mapper, context, ksessionId, getIdentifier()));
-                    }
-                } else {
-                    ksession = factory.findKieSessionById(ksessionId);
-                }
-            }
+        RuntimeEngine localRuntime = findLocalRuntime(contextId);
+        if (localRuntime != null) {
+            logger.debug("Runtime engine found in local {} with context {}", localRuntime, contextId);
+            return localRuntime;
+        }
+
+        Long ksessionId = mapper.findMapping(context, this.getIdentifier());
+        if(ksessionId == null && mapper instanceof InternalMapper) {
+            // we fall back as process instance must be closed but not really case and kie session must be still there.
+            logger.debug("PerCaseRuntimeManager KieSessionId found fallingback for context {} and owner {}", context.getContextId(), getIdentifier());
+            ksessionId = ((InternalMapper) mapper).findLogMapping(context, identifier);
```

Comment:
Same here, out of the synchronized block.

Reply:
The integrity of the mapping existence is controlled by DB so there is no need for this to be in a sync block. Any operation adding / removing the mapping is atomic to the transaction so that is not going to change during this tx no matter how many calls you do.
The sync block in this case does not make much sense.